### PR TITLE
feat: ssr middlware

### LINKED ISSUE #66: ssr middleware
https://twitter.com/dai_shi/status/1664804156501098497

I will make `ssr` middleware totally isolated from `rsc` middlware (assuming they work on different machines). Not sure how it goes yet.
Refactoring builder would be required in advance, and we might need some breaking change for `getBuilder` API.

### PR DESCRIPTION
**Pull request by dai-shi**
Close #66 

- [x] first experiment
- [x] dev server with examples/01_counter
- [x] dev server with examples/07_router
- [x] dev server with examples/* (except for 06_nesting)
- [x] prd server
- [x] build and generate static sites

https://excalidraw.com/#json=pSsBQOqkYX4O-TIgxNrZj,i_CsymP1VZiHIAa0dlLbNw

image

### CODE REVIEW

**Comment by dai-shi** [comment]
Found a relevant comment: https://github.com/facebook/react/issues/25627#issuecomment-1444073227

> You can suppress the error you've thrown from the console in onRecoverableError handler in your hydrateRoot options.

Let's do that.

File: src/lib/middleware/ssr.ts [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+      }
+      if (!htmlRes.body || !rscRes.body) {
+        passthrough.destroy(new Error("No body"));
+        return;
+      }
+      const htmlStr = await htmlRes.text(); // Hope stream works, but it'd be too tricky
+      const [preamble, postamble] = splitHTML(htmlStr);
+      passthrough.write(preamble, "utf8");
+      const data = createFromNodeStream(
+        Readable.fromWeb(rscRes.body as any), // FIXME how to avoid any?
```
Comment (Himself65) [inline src/lib/middleware/ssr.ts:91]: Update to `"@types/node": "^20.3.1"` could fix this issue

This is upstream bug

**Review by Himself65 — CHANGES_REQUESTED** [review]

File: src/lib/config.ts [hunk 2]
```diff
@@ -2,18 +2,49 @@ import { resolveConfig as viteResolveConfig } from "vite";
...
+  }
+  return [splitted[0] + startStr, endStr + splitted[1]];
+};
+
+const getFallback = (id: string) => {
+  if (id.endsWith("#Waku_SSR_Capable_Link")) {
+    return "waku/server#ClientOnly";
+  }
+  return "waku/server#ClientFallback";
+};
```
Comment (dai-shi) [inline src/lib/config.ts:26]: I tried to create a full SSR (= also render client components into HTML on server) example, with configuring `getFallback`, but there are a few issues and I gave it up doing in this PR:
- We need a module graph in DEV, which we don't have in our current architecture.
- In production mode, it should be easier, but the configuration would be tricker.
- What's really blocking is probably as we load React in a separate worker, we get "Invalid hook call." error. I don't think we can do SSR in the same worker, so we'd need a new separate worker and duplicate the work. That doesn't feel very RSC-first framework.

We will see it again in the future, maybe after we migrate to `react-server-dom-vite`, which is not yet planned.

File: examples/02_async/entries.ts [hunk 3]
```diff
@@ -18,4 +18,14 @@ export default defineEntries(
       },
     };
   }
+  // getSsrConfig
+  // FIXME SSR streaming and RSC streaming conflict. Let's revisit it later.
```
Comment (dai-shi) [inline examples/02_async/entries.ts:22]: I think I understand what's happening. We need to send our `<script>` in the middle of SSR streaming. Let me see how we could do that in a simple way... 🤔

Comment (dai-shi) [reply examples/02_async/entries.ts:22]: #78

**Pull request merged**